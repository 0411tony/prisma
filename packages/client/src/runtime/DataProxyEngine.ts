import {
  Engine,
  EngineConfig,
  EngineEventType,
  GetConfigResult,
} from '@prisma/engine-core/src/common/Engine'
import fs from 'fs'
import { createHash } from 'crypto'
import { URL } from 'url'
import EventEmitter from 'events'
import * as prismafile from 'prismafile'
import fetch from 'node-fetch'

const BACKOFF_INTERVAL = 250
const MAX_RETRIES = 5

function backOff(n: number): Promise<number> {
  const baseDelay = Math.pow(2, n) * BACKOFF_INTERVAL
  const jitter = Math.ceil(Math.random() * baseDelay) - Math.ceil(baseDelay / 2)
  const total = baseDelay + jitter

  return new Promise((done) => setTimeout(() => done(total), total))
}

// TODO: move to @prisma/engine-core
export class DataProxyEngine extends Engine {
  private schemaText: string
  private schemaBase64: string
  private schemaHash: string

  private url: (s: string) => string
  private headers: { Authorization: string }

  private logEmitter = new EventEmitter()

  constructor(private config: EngineConfig) {
    super()

    this.schemaText = fs.readFileSync(config.datamodelPath, 'utf8')
    this.schemaBase64 = Buffer.from(this.schemaText).toString('base64')
    this.schemaHash = createHash('sha256')
      .update(this.schemaBase64)
      .digest('hex')

    const [host, apiKey] = extractHostAndApiKey(this.schemaText)
    // TODO: how to get the current version the right way?
    const clientVersion = '2.29.1'

    this.url = (s) => `https://${host}/${clientVersion}/${this.schemaHash}/${s}`
    this.headers = { Authorization: `Bearer ${apiKey}` }

    this.logEmitter.on('error', () => {
      // Prevent unhandled error events
    })

    this.logEmitter.emit('info', `Ready for ${host}`)
  }

  version() {
    // QE is remote, we don't need to know the exact commit SHA
    return 'unknown'
  }

  async start() {}
  async stop() {}

  on(event: EngineEventType, listener: (args?: any) => any): void {
    if (event === 'beforeExit') {
      // TODO: hook into the process
      throw new Error('beforeExit event is not supported yet')
    } else {
      this.logEmitter.on(event, listener)
    }
  }

  // TODO: looks like activeProvider is the only thing
  // used externally; verify that
  getConfig() {
    return Promise.resolve({
      datasources: [
        {
          activeProvider: this.config.activeProvider,
        },
      ],
    } as GetConfigResult)
  }

  private async uploadSchema() {
    const res = await fetch(this.url('schema'), {
      method: 'PUT',
      headers: this.headers,
      body: this.schemaBase64,
    })

    if (!res.ok) {
      this.logEmitter.emit('warn', 'Could not upload the schema')
      throw new Error('Could not upload the schema')
    }
  }

  async request<T>(q: string, _hs?: any, n = 0) {
    try {
      this.logEmitter.emit('query', q)

      const res = await fetch(this.url('graphql'), {
        method: 'POST',
        headers: this.headers,
        body: JSON.stringify({
          operationName: null,
          variables: {},
          query: q,
        }),
      })

      // 404 on the GraphQL endpoint may mean that the schema
      // was not uploaded yet.
      if (res.status === 404) {
        await this.uploadSchema()
        throw new Error('Schema (re)uploaded')
      }

      if (!res.ok) {
        throw new Error('GraphQL request failed')
      }

      return JSON.parse(await res.text())
    } catch (err) {
      if (n >= MAX_RETRIES) {
        this.logEmitter.emit('error', `Failed to query: ${err.message}`)
        throw err
      } else {
        const delay = await backOff(n)
        this.logEmitter.emit('warn', `Retrying after ${delay}`)
        return this.request<T>(q, {}, n + 1)
      }
    }
  }

  requestBatch<T>(qs: string[], _hs: any, transact = false) {
    // TODO: transactions not supported (see below)
    // TODO: introduce batch endpoint in Data Proxy;
    // current implementation will result in `qs.length` HTTP calls
    return transact
      ? this.transaction()
      : Promise.all(qs.map((q) => this.request<T>(q)))
  }

  // TODO: figure out how to support transactions
  transaction(): Promise<any> {
    throw new Error('Transactions are currently not supported in Data Proxy')
  }
}

// Regex + prismafile feels fragile...
// TODO: build a minimal single-purpose (URL extraction)
// and well-tested TS schema parser
function extractHostAndApiKey(schemaText: string) {
  // Extract the datasource block so we feed prismafile with minimal input
  const strippedToDatasource = /datasource[ \t]+[^\s]+[ \t]+\{[^}]+}/.exec(
    schemaText,
  )
  if (!strippedToDatasource) {
    throw new Error('Could not find a valid datasource block in the schema')
  }

  const datasourceText = strippedToDatasource[0]
  if (typeof datasourceText !== 'string' || datasourceText.length < 1) {
    throw new Error('Could not find a valid datasource block in the schema')
  }

  // Use prismafile to determine the URL
  let url = ''
  try {
    const schemaAst = prismafile.parse(datasourceText)
    const [datasource] = schemaAst.blocks.filter(
      (b) => b.type === 'datasource',
    ) as any[]

    // Datasource type is not exported from prismafile, so from now on
    // we're in any-land. Hence the dirtiness and the big try/catch around.

    const urlAssignment = datasource.assignments.find(
      (a: any) => a.key?.type === 'identifier' && a.key?.name === 'url',
    )

    if (!urlAssignment || !urlAssignment.value) {
      throw new Error('No assignment found')
    }

    const { type, value, name, arguments: args } = urlAssignment.value

    if (type === 'string_value' && typeof value === 'string') {
      url = value // Datasource URL directly in the schema
    } else if (
      type === 'function_value' &&
      name?.type === 'identifier' &&
      name.name === 'env' &&
      Array.isArray(args)
    ) {
      const arg = args[0]

      if (arg?.type === 'string_value' && typeof arg.value === 'string') {
        const envVarName = arg.value.trim()

        if (envVarName.length > 0) {
          // Datasource URL in environment
          url = process.env[envVarName] ?? ''
        }
      }
    }

    url = url.trim()
    if (url.length < 1) {
      throw new Error('No URL in schema nor environment')
    }
  } catch (err) {
    throw new Error('Could not extract URL of the datasource')
  }

  let parsed: URL
  try {
    parsed = new URL(url)
  } catch (err) {
    throw new Error('Could not parse URL of the datasource')
  }

  const { protocol, host, searchParams } = parsed

  if (protocol !== 'prisma:') {
    throw new Error('Datasource URL should use prisma:// protocol')
  }

  if (host.length < 1) {
    throw new Error('No valid host found in the datasource URL')
  }

  const apiKey = searchParams.get('api_key')
  if (!apiKey || apiKey.length < 1) {
    throw new Error('No valid API key found in the datasource URL')
  }

  return [host, apiKey]
}
