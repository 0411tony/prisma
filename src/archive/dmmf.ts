export const dmmf = {
  datamodel: {
    models: [
      {
        name: 'User',
        isEmbedded: false,
        isEnum: false,
        dbName: '',
        fields: [
          {
            kind: 'scalar',
            name: 'id',
            arity: 'required',
            isUnique: true,
            isId: true,
            type: 'ID',
          },
          {
            kind: 'scalar',
            name: 'name',
            arity: 'required',
            isUnique: false,
            isId: false,
            type: 'String',
          },
          {
            kind: 'scalar',
            name: 'strings',
            arity: 'list',
            isUnique: false,
            isId: false,
            type: 'String',
          },
          {
            kind: 'relation',
            name: 'posts',
            arity: 'list',
            isUnique: false,
            isId: false,
            type: 'Post',
          },
        ],
      },
      {
        name: 'Post',
        isEmbedded: false,
        isEnum: false,
        dbName: '',
        fields: [
          {
            kind: 'scalar',
            name: 'id',
            arity: 'required',
            isUnique: true,
            isId: true,
            type: 'ID',
          },
          {
            kind: 'scalar',
            name: 'title',
            arity: 'required',
            isUnique: false,
            isId: false,
            type: 'String',
          },
          {
            kind: 'scalar',
            name: 'content',
            arity: 'required',
            isUnique: false,
            isId: false,
            type: 'String',
          },
          {
            kind: 'relation',
            name: 'author',
            arity: 'required',
            isUnique: false,
            isId: false,
            type: 'User',
          },
        ],
      },
    ],
  },
  schema: {
    queries: [
      {
        name: 'users',
        args: [
          {
            name: 'where',
            type: 'UserWhereInput',
            arity: 'optional',
          },
          {
            name: 'orderBy',
            type: 'UserOrderByInput',
            arity: 'optional',
          },
          {
            name: 'skip',
            type: 'Int',
            arity: 'optional',
          },
          {
            name: 'after',
            type: 'String',
            arity: 'optional',
          },
          {
            name: 'before',
            type: 'String',
            arity: 'optional',
          },
          {
            name: 'first',
            type: 'Int',
            arity: 'optional',
          },
          {
            name: 'last',
            type: 'Int',
            arity: 'optional',
          },
        ],
        output: {
          name: 'User',
          arity: 'list',
        },
      },
      {
        name: 'posts',
        args: [
          {
            name: 'where',
            type: 'PostWhereInput',
            arity: 'optional',
          },
          {
            name: 'orderBy',
            type: 'PostOrderByInput',
            arity: 'optional',
          },
          {
            name: 'skip',
            type: 'Int',
            arity: 'optional',
          },
          {
            name: 'after',
            type: 'String',
            arity: 'optional',
          },
          {
            name: 'before',
            type: 'String',
            arity: 'optional',
          },
          {
            name: 'first',
            type: 'Int',
            arity: 'optional',
          },
          {
            name: 'last',
            type: 'Int',
            arity: 'optional',
          },
        ],
        output: {
          name: 'Post',
          arity: 'list',
        },
      },
      {
        name: 'user',
        args: [
          {
            name: 'where',
            type: 'UserWhereUniqueInput',
            arity: 'required',
          },
        ],
        output: {
          name: 'User',
          arity: 'optional',
        },
      },
      {
        name: 'post',
        args: [
          {
            name: 'where',
            type: 'PostWhereUniqueInput',
            arity: 'required',
          },
        ],
        output: {
          name: 'Post',
          arity: 'optional',
        },
      },
      {
        name: 'usersConnection',
        args: [
          {
            name: 'where',
            type: 'UserWhereInput',
            arity: 'optional',
          },
          {
            name: 'orderBy',
            type: 'UserOrderByInput',
            arity: 'optional',
          },
          {
            name: 'skip',
            type: 'Int',
            arity: 'optional',
          },
          {
            name: 'after',
            type: 'String',
            arity: 'optional',
          },
          {
            name: 'before',
            type: 'String',
            arity: 'optional',
          },
          {
            name: 'first',
            type: 'Int',
            arity: 'optional',
          },
          {
            name: 'last',
            type: 'Int',
            arity: 'optional',
          },
        ],
        output: {
          name: 'UserConnection',
          arity: 'required',
        },
      },
      {
        name: 'postsConnection',
        args: [
          {
            name: 'where',
            type: 'PostWhereInput',
            arity: 'optional',
          },
          {
            name: 'orderBy',
            type: 'PostOrderByInput',
            arity: 'optional',
          },
          {
            name: 'skip',
            type: 'Int',
            arity: 'optional',
          },
          {
            name: 'after',
            type: 'String',
            arity: 'optional',
          },
          {
            name: 'before',
            type: 'String',
            arity: 'optional',
          },
          {
            name: 'first',
            type: 'Int',
            arity: 'optional',
          },
          {
            name: 'last',
            type: 'Int',
            arity: 'optional',
          },
        ],
        output: {
          name: 'PostConnection',
          arity: 'required',
        },
      },
      {
        name: 'node',
        args: [
          {
            name: 'id',
            type: 'ID',
            arity: 'required',
          },
        ],
        output: {
          name: 'Node',
          arity: 'optional',
        },
      },
    ],
    mutations: [
      {
        name: 'createUser',
        args: [
          {
            name: 'data',
            type: 'UserCreateInput',
            arity: 'required',
          },
        ],
        output: {
          name: 'User',
          arity: 'required',
        },
      },
      {
        name: 'createPost',
        args: [
          {
            name: 'data',
            type: 'PostCreateInput',
            arity: 'required',
          },
        ],
        output: {
          name: 'Post',
          arity: 'required',
        },
      },
      {
        name: 'updateUser',
        args: [
          {
            name: 'data',
            type: 'UserUpdateInput',
            arity: 'required',
          },
          {
            name: 'where',
            type: 'UserWhereUniqueInput',
            arity: 'required',
          },
        ],
        output: {
          name: 'User',
          arity: 'optional',
        },
      },
      {
        name: 'updatePost',
        args: [
          {
            name: 'data',
            type: 'PostUpdateInput',
            arity: 'required',
          },
          {
            name: 'where',
            type: 'PostWhereUniqueInput',
            arity: 'required',
          },
        ],
        output: {
          name: 'Post',
          arity: 'optional',
        },
      },
      {
        name: 'deleteUser',
        args: [
          {
            name: 'where',
            type: 'UserWhereUniqueInput',
            arity: 'required',
          },
        ],
        output: {
          name: 'User',
          arity: 'optional',
        },
      },
      {
        name: 'deletePost',
        args: [
          {
            name: 'where',
            type: 'PostWhereUniqueInput',
            arity: 'required',
          },
        ],
        output: {
          name: 'Post',
          arity: 'optional',
        },
      },
      {
        name: 'upsertUser',
        args: [
          {
            name: 'where',
            type: 'UserWhereUniqueInput',
            arity: 'required',
          },
          {
            name: 'create',
            type: 'UserCreateInput',
            arity: 'required',
          },
          {
            name: 'update',
            type: 'UserUpdateInput',
            arity: 'required',
          },
        ],
        output: {
          name: 'User',
          arity: 'required',
        },
      },
      {
        name: 'upsertPost',
        args: [
          {
            name: 'where',
            type: 'PostWhereUniqueInput',
            arity: 'required',
          },
          {
            name: 'create',
            type: 'PostCreateInput',
            arity: 'required',
          },
          {
            name: 'update',
            type: 'PostUpdateInput',
            arity: 'required',
          },
        ],
        output: {
          name: 'Post',
          arity: 'required',
        },
      },
      {
        name: 'updateManyUsers',
        args: [
          {
            name: 'data',
            type: 'UserUpdateManyMutationInput',
            arity: 'required',
          },
          {
            name: 'where',
            type: 'UserWhereInput',
            arity: 'optional',
          },
        ],
        output: {
          name: 'BatchPayload',
          arity: 'required',
        },
      },
      {
        name: 'updateManyPosts',
        args: [
          {
            name: 'data',
            type: 'PostUpdateManyMutationInput',
            arity: 'required',
          },
          {
            name: 'where',
            type: 'PostWhereInput',
            arity: 'optional',
          },
        ],
        output: {
          name: 'BatchPayload',
          arity: 'required',
        },
      },
      {
        name: 'deleteManyUsers',
        args: [
          {
            name: 'where',
            type: 'UserWhereInput',
            arity: 'optional',
          },
        ],
        output: {
          name: 'BatchPayload',
          arity: 'required',
        },
      },
      {
        name: 'deleteManyPosts',
        args: [
          {
            name: 'where',
            type: 'PostWhereInput',
            arity: 'optional',
          },
        ],
        output: {
          name: 'BatchPayload',
          arity: 'required',
        },
      },
    ],
    inputTypes: [
      {
        name: 'UserWhereInput',
        args: [
          {
            name: 'AND',
            type: 'UserWhereInput',
          },
          {
            name: 'id',
            type: 'ID',
          },
          {
            name: 'id_not',
            type: 'ID',
          },
          {
            name: 'id_in',
            type: 'ID',
          },
          {
            name: 'id_not_in',
            type: 'ID',
          },
          {
            name: 'id_lt',
            type: 'ID',
          },
          {
            name: 'id_lte',
            type: 'ID',
          },
          {
            name: 'id_gt',
            type: 'ID',
          },
          {
            name: 'id_gte',
            type: 'ID',
          },
          {
            name: 'id_contains',
            type: 'ID',
          },
          {
            name: 'id_not_contains',
            type: 'ID',
          },
          {
            name: 'id_starts_with',
            type: 'ID',
          },
          {
            name: 'id_not_starts_with',
            type: 'ID',
          },
          {
            name: 'id_ends_with',
            type: 'ID',
          },
          {
            name: 'id_not_ends_with',
            type: 'ID',
          },
          {
            name: 'name',
            type: 'String',
          },
          {
            name: 'name_not',
            type: 'String',
          },
          {
            name: 'name_in',
            type: 'String',
          },
          {
            name: 'name_not_in',
            type: 'String',
          },
          {
            name: 'name_lt',
            type: 'String',
          },
          {
            name: 'name_lte',
            type: 'String',
          },
          {
            name: 'name_gt',
            type: 'String',
          },
          {
            name: 'name_gte',
            type: 'String',
          },
          {
            name: 'name_contains',
            type: 'String',
          },
          {
            name: 'name_not_contains',
            type: 'String',
          },
          {
            name: 'name_starts_with',
            type: 'String',
          },
          {
            name: 'name_not_starts_with',
            type: 'String',
          },
          {
            name: 'name_ends_with',
            type: 'String',
          },
          {
            name: 'name_not_ends_with',
            type: 'String',
          },
          {
            name: 'posts_some',
            type: 'PostWhereInput',
          },
        ],
      },
      {
        name: 'UserOrderByInput',
        args: [
          {
            name: 'id_ASC',
            type: 'UserOrderByInput',
          },
          {
            name: 'id_DESC',
            type: 'UserOrderByInput',
          },
          {
            name: 'name_ASC',
            type: 'UserOrderByInput',
          },
          {
            name: 'name_DESC',
            type: 'UserOrderByInput',
          },
        ],
      },
      {
        name: 'PostWhereInput',
        args: [
          {
            name: 'AND',
            type: 'PostWhereInput',
          },
          {
            name: 'id',
            type: 'ID',
          },
          {
            name: 'id_not',
            type: 'ID',
          },
          {
            name: 'id_in',
            type: 'ID',
          },
          {
            name: 'id_not_in',
            type: 'ID',
          },
          {
            name: 'id_lt',
            type: 'ID',
          },
          {
            name: 'id_lte',
            type: 'ID',
          },
          {
            name: 'id_gt',
            type: 'ID',
          },
          {
            name: 'id_gte',
            type: 'ID',
          },
          {
            name: 'id_contains',
            type: 'ID',
          },
          {
            name: 'id_not_contains',
            type: 'ID',
          },
          {
            name: 'id_starts_with',
            type: 'ID',
          },
          {
            name: 'id_not_starts_with',
            type: 'ID',
          },
          {
            name: 'id_ends_with',
            type: 'ID',
          },
          {
            name: 'id_not_ends_with',
            type: 'ID',
          },
          {
            name: 'title',
            type: 'String',
          },
          {
            name: 'title_not',
            type: 'String',
          },
          {
            name: 'title_in',
            type: 'String',
          },
          {
            name: 'title_not_in',
            type: 'String',
          },
          {
            name: 'title_lt',
            type: 'String',
          },
          {
            name: 'title_lte',
            type: 'String',
          },
          {
            name: 'title_gt',
            type: 'String',
          },
          {
            name: 'title_gte',
            type: 'String',
          },
          {
            name: 'title_contains',
            type: 'String',
          },
          {
            name: 'title_not_contains',
            type: 'String',
          },
          {
            name: 'title_starts_with',
            type: 'String',
          },
          {
            name: 'title_not_starts_with',
            type: 'String',
          },
          {
            name: 'title_ends_with',
            type: 'String',
          },
          {
            name: 'title_not_ends_with',
            type: 'String',
          },
          {
            name: 'content',
            type: 'String',
          },
          {
            name: 'content_not',
            type: 'String',
          },
          {
            name: 'content_in',
            type: 'String',
          },
          {
            name: 'content_not_in',
            type: 'String',
          },
          {
            name: 'content_lt',
            type: 'String',
          },
          {
            name: 'content_lte',
            type: 'String',
          },
          {
            name: 'content_gt',
            type: 'String',
          },
          {
            name: 'content_gte',
            type: 'String',
          },
          {
            name: 'content_contains',
            type: 'String',
          },
          {
            name: 'content_not_contains',
            type: 'String',
          },
          {
            name: 'content_starts_with',
            type: 'String',
          },
          {
            name: 'content_not_starts_with',
            type: 'String',
          },
          {
            name: 'content_ends_with',
            type: 'String',
          },
          {
            name: 'content_not_ends_with',
            type: 'String',
          },
          {
            name: 'author',
            type: 'UserWhereInput',
          },
        ],
      },
      {
        name: 'PostOrderByInput',
        args: [
          {
            name: 'id_ASC',
            type: 'PostOrderByInput',
          },
          {
            name: 'id_DESC',
            type: 'PostOrderByInput',
          },
          {
            name: 'title_ASC',
            type: 'PostOrderByInput',
          },
          {
            name: 'title_DESC',
            type: 'PostOrderByInput',
          },
          {
            name: 'content_ASC',
            type: 'PostOrderByInput',
          },
          {
            name: 'content_DESC',
            type: 'PostOrderByInput',
          },
        ],
      },
      {
        name: 'UserWhereUniqueInput',
        args: [
          {
            name: 'id',
            type: 'ID',
          },
        ],
      },
      {
        name: 'PostWhereUniqueInput',
        args: [
          {
            name: 'id',
            type: 'ID',
          },
        ],
      },
      {
        name: 'UserCreateInput',
        args: [
          {
            name: 'id',
            type: 'ID',
          },
          {
            name: 'name',
            type: 'String',
          },
          {
            name: 'strings',
            type: 'UserCreatestringsInput',
          },
          {
            name: 'posts',
            type: 'PostCreateManyWithoutAuthorInput',
          },
        ],
      },
      {
        name: 'PostCreateInput',
        args: [
          {
            name: 'id',
            type: 'ID',
          },
          {
            name: 'title',
            type: 'String',
          },
          {
            name: 'content',
            type: 'String',
          },
          {
            name: 'author',
            type: 'UserCreateOneWithoutPostsInput',
          },
        ],
      },
      {
        name: 'UserUpdateInput',
        args: [
          {
            name: 'name',
            type: 'String',
          },
          {
            name: 'strings',
            type: 'UserUpdatestringsInput',
          },
          {
            name: 'posts',
            type: 'PostUpdateManyWithoutAuthorInput',
          },
        ],
      },
      {
        name: 'PostUpdateInput',
        args: [
          {
            name: 'title',
            type: 'String',
          },
          {
            name: 'content',
            type: 'String',
          },
          {
            name: 'author',
            type: 'UserUpdateOneRequiredWithoutPostsInput',
          },
        ],
      },
      {
        name: 'UserUpdateManyMutationInput',
        args: [
          {
            name: 'name',
            type: 'String',
          },
          {
            name: 'strings',
            type: 'UserUpdatestringsInput',
          },
        ],
      },
      {
        name: 'PostUpdateManyMutationInput',
        args: [
          {
            name: 'title',
            type: 'String',
          },
          {
            name: 'content',
            type: 'String',
          },
        ],
      },
    ],
    outputTypes: [
      {
        name: 'User',
        fields: [
          {
            name: 'id',
            type: 'ID',
            arity: 'required',
          },
          {
            name: 'name',
            type: 'String',
            arity: 'required',
          },
          {
            name: 'strings',
            arity: 'list',
          },
          {
            name: 'posts',
            type: 'Post',
            arity: 'optional',
          },
        ],
      },
      {
        name: 'Post',
        fields: [
          {
            name: 'id',
            type: 'ID',
            arity: 'required',
          },
          {
            name: 'title',
            type: 'String',
            arity: 'required',
          },
          {
            name: 'content',
            type: 'String',
            arity: 'required',
          },
          {
            name: 'author',
            type: 'User',
            arity: 'required',
          },
        ],
      },
      {
        name: 'UserConnection',
        fields: [
          {
            name: 'pageInfo',
            type: 'PageInfo',
            arity: 'required',
          },
          {
            name: 'edges',
            type: 'UserEdge',
            arity: 'list',
          },
          {
            name: 'aggregate',
            type: 'AggregateUser',
            arity: 'required',
          },
        ],
      },
      {
        name: 'PostConnection',
        fields: [
          {
            name: 'pageInfo',
            type: 'PageInfo',
            arity: 'required',
          },
          {
            name: 'edges',
            type: 'PostEdge',
            arity: 'list',
          },
          {
            name: 'aggregate',
            type: 'AggregatePost',
            arity: 'required',
          },
        ],
      },
      {
        name: 'Node',
        fields: [
          {
            name: 'id',
            type: 'ID',
            arity: 'required',
          },
        ],
      },
      {
        name: 'BatchPayload',
        fields: [
          {
            name: 'count',
            type: 'Long',
            arity: 'required',
          },
        ],
      },
    ],
  },
  mappings: [
    {
      model: 'User',
      findOne: 'user',
      findMany: 'users',
      create: 'createUser',
      update: 'updateUser',
      updateMany: 'updateManyUser',
      upsert: 'upsertUser',
      delete: 'deleteUser',
      deleteMany: 'deleteManyUser',
    },
    {
      model: 'Post',
      findOne: 'post',
      findMany: 'posts',
      create: 'createPost',
      update: 'updatePost',
      updateMany: 'updateManyPost',
      upsert: 'upsertPost',
      delete: 'deletePost',
      deleteMany: 'deleteManyPost',
    },
  ],
}
