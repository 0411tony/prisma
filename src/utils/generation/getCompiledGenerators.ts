import { CompiledGeneratorDefinition, Dictionary, GeneratorDefinitionWithPackage } from '@prisma/cli'
import path from 'path'
import { LiftEngine } from '../../LiftEngine'
import chalk from 'chalk'
import fs from 'fs-extra'
import { runGeneratorBinary } from './runGeneratorBinary'
// import { generateInThread } from '../generateInThread'

const didYouMeanMap = {
  javascript: 'photonjs',
  typescript: 'photonjs',
  'photon-js': 'photonjs',
  photon: 'photonjs',
}

export async function getCompiledGenerators(
  cwd: string,
  datamodel: string,
  definitions: Dictionary<GeneratorDefinitionWithPackage>,
): Promise<CompiledGeneratorDefinition[]> {
  const engine = new LiftEngine({ projectDir: cwd })
  const config = await engine.getConfig({ datamodel })

  const generators = config.generators
    .slice()
    .sort((a, b) => (a.provider === 'photonjs' ? -1 : b.provider === 'photonjs' ? 1 : a.name < b.name ? -1 : 1))

  return config.generators.map((g, index) => {
    const otherGenerators = generators.filter((g, i) => i !== index)
    if (g.provider.startsWith('.') || g.provider.startsWith('/')) {
      const binPath = path.resolve(cwd, g.provider)
      if (!fs.pathExistsSync(binPath)) {
        throw new Error(`Could not find generator binary ${chalk.bold(g.provider)}`)
      }
      if (!hasChmodX(binPath)) {
        throw new Error(
          `${chalk.bold(g.provider)} is not executable. Please run ${chalk.greenBright(`chmod +x ${g.provider}`)}`,
        )
      }
      return {
        prettyName: binPath,
        generate: async () =>
          (await runGeneratorBinary(binPath, {
            cwd,
            generator: {
              ...g,
              output: g.output ? path.resolve(process.cwd(), g.output) : undefined,
            },
            otherGenerators,
          })).stdout,
      }
    }

    const predefinedGenerator = definitions[g.provider]
    if (!predefinedGenerator) {
      const replacement = didYouMeanMap[g.provider]
      const didYouMean = replacement ? `\nDid you mean ${chalk.greenBright(replacement)}?` : ''
      throw new Error(
        `Unknown generator provider ${g.provider} for generator ${g.name} defined in ${chalk.underline(
          path.join(cwd, 'project.prisma'),
        )}${didYouMean}`,
      )
    }
    const output =
      g.output ||
      predefinedGenerator.definition.defaultOutput ||
      `node_modules/@generated/${predefinedGenerator.packagePath}`

    return {
      prettyName: predefinedGenerator.definition.prettyName,
      generate: () =>
        predefinedGenerator.definition.generate({
          cwd,
          generator: { ...g, output },
          otherGenerators,
        }),
    } as CompiledGeneratorDefinition
  })
}

function hasChmodX(file: string): boolean {
  const s = fs.statSync(file)
  const newMode = s.mode | 64 | 8 | 1
  return s.mode === newMode
}
