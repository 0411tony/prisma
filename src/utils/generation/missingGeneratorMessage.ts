import chalk from 'chalk'
import { highlightDatamodel } from '../../cli/highlight/highlight'

export const missingGeneratorMessage = `\n${chalk.blue(
  'info',
)} You don't have defined any generator in your ${chalk.bold(
  'schema.prisma',
)}, so nothing will be generated. You can define them like this:

${chalk.bold(
  highlightDatamodel(`generator client {
  provider = "prisma-client-js"
}`),
)}`
