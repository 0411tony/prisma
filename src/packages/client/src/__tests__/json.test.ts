import { DMMFClass } from '../runtime/dmmf'
import { makeDocument } from '../runtime/query'
import { getDMMF } from '../generation/getDMMF'

const datamodel = `\
datasource db {
  provider = "postgres"
  url = env("DB")
}

model User {
  id    Int @id @default(autoincrement())
  name  String
  email String @unique
  json  Json
}`

let dmmf
beforeAll(async () => {
  const dmmfDocument = await getDMMF({ datamodel })
  dmmf = new DMMFClass(dmmfDocument)
})

describe('json', () => {
  test('should be able to create json', async () => {
    const document = makeDocument({
      dmmf,
      select: {
        data: {
          email: 'a@a.de',
          json: {
            hello: 'world',
          },
          name: 'Bob',
        },
      },
      rootTypeName: 'mutation',
      rootField: 'createOneUser',
    })
    expect(String(document)).toMatchInlineSnapshot(`
      "mutation {
        createOneUser(data: {
          email: \\"a@a.de\\"
          json: \\"{\\\\\\"hello\\\\\\":\\\\\\"world\\\\\\"}\\"
          name: \\"Bob\\"
        }) {
          id
          name
          email
          json
        }
      }"
    `)
  })
  test('should reject directly querying json with the equals shortcut', () => {
    const select = {
      where: {
        json: {
          hello: 'world',
        },
      },
    }
    const document = makeDocument({
      dmmf,
      select,
      rootTypeName: 'query',
      rootField: 'findManyUser',
    })
    expect(() =>
      document.validate(select, false, 'user.findMany', 'colorless'),
    ).toThrowErrorMatchingSnapshot()
    expect(String(document)).toMatchInlineSnapshot(`
      "query {
        findManyUser(where: {
          json: {
            hello: \\"world\\"
          }
        }) {
          id
          name
          email
          json
        }
      }"
    `)
  })
  test('should be able to query json with equals', () => {
    const select = {
      where: {
        json: {
          equals: {
            hello: 'world',
          },
        },
      },
    }
    const document = makeDocument({
      dmmf,
      select,
      rootTypeName: 'query',
      rootField: 'findManyUser',
    })
    document.validate(select)
    expect(String(document)).toMatchInlineSnapshot(`
      "query {
        findManyUser(where: {
          json: {
            equals: \\"{\\\\\\"hello\\\\\\":\\\\\\"world\\\\\\"}\\"
          }
        }) {
          id
          name
          email
          json
        }
      }"
    `)
  })
  test('should be able to query json with not', () => {
    const select = {
      where: {
        json: {
          not: {
            hello: 'world',
          },
        },
      },
    }
    const document = makeDocument({
      dmmf,
      select,
      rootTypeName: 'query',
      rootField: 'findManyUser',
    })
    document.validate(select)
    expect(String(document)).toMatchInlineSnapshot(`
      "query {
        findManyUser(where: {
          json: {
            not: \\"{\\\\\\"hello\\\\\\":\\\\\\"world\\\\\\"}\\"
          }
        }) {
          id
          name
          email
          json
        }
      }"
    `)
  })
})
