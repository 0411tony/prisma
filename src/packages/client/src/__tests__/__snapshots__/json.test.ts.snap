// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json should reject directly querying json with the equals shortcut 1`] = `
"
Invalid \`prisma.user.findMany()\` invocation:

{
  where: {
    json: {
      hello: 'world'
      ~~~~~
    }
  }
}

Unknown arg \`hello\` in where.json.hello for type JsonFilter. Did you mean \`select\`? Available args:
type JsonFilter {
  equals?: Json
  not?: Json | JsonFilter
}

"
`;
